server/src/internalClusterTest/java/org/elasticsearch/recovery/TruncatedRecoveryIT.java

```diff
@@ -104,18 +105,43 @@ public void testCancelRecoveryAndResume() throws Exception {
                 .addSendBehavior(
-                    internalCluster().getInstance(TransportService.class, unluckyNode.getNode().getName()),
+                    internalCluster().getInstance(TransportService.class, unluckyNode),
                     (connection, requestId, action, request, options) -> {
                         if (action.equals(PeerRecoveryTargetService.Actions.FILE_CHUNK)) {
                             RecoveryFileChunkRequest req = (RecoveryFileChunkRequest) request;
                             logger.info("file chunk [{}] lastChunk: {}", req, req.lastChunk());
                             if ((req.name().endsWith("cfs") || req.name().endsWith("fdt")) && req.lastChunk() && truncate.get()) {
+                                // write some garbage in the shard directory so we can verify that it is cleaned up before we resend.
+                                // This helps prevent recovery from failing due to lack of space from garbage left over from a previous
+                                // recovery that crashed during file transmission.
+                                // We can't look for the presence of the recovery temp files themselves because they are automatically
+                                // cleaned up on clean shutdown by MultiFileWriter.
+                                var shardId = req.shardId();
+                                var shardPath = unluckyIndices.indexService(shardId.getIndex())
+                                    .getShard(shardId.getId())
+                                    .shardPath()
+                                    .resolveIndex();
+                                var garbagePath = Files.createFile(shardPath.resolve("recovery.garbage"));
+                                logger.info("writing garbage at: {}", garbagePath);
                                 latch.countDown();
                                 throw new RuntimeException("Caused some truncated files for fun and profit");
                             }
+                        } else if (action.equals(PeerRecoveryTargetService.Actions.FILES_INFO) && truncate.get() == false) {
+                            // verify on the second recovery attempt that the garbage has been removed before we fetch files again #104473
+                            var shardId = ((RecoveryFilesInfoRequest) request).shardId();
+                            var shardPath = unluckyIndices.indexService(shardId.getIndex())
+                                .getShard(shardId.getId())
+                                .shardPath()
+                                .resolveIndex();
+                            logger.info("checking for garbage at {}", shardPath.resolve("recovery.garbage"));
+                            assertFalse(
+                                "garbage file should have been cleaned before file transmission",
+                                Files.exists(shardPath.resolve("garbage"))
```

Comment:
I believe this checks for a wrong file, not the one created earlier (`recovery.garbage` vs `garbage`).
It would be nice to extract its name as variable earlier in the code and possibly randomize the file name.

Reply:
Whoops! Sounds good, I'll randomly generate some during round 1 and verify that they are all gone before transfer in round 2.